# Imputation using Regression

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

In [2]:
df=pd.read_csv('framingham.csv')
print(df.shape)
df.head()

(4238, 16)


sex  age  education  currentSmoker  cigsPerDay  BPMeds  prevalentStroke  \
0    1   39        4.0              0         0.0     0.0                0   
1    0   46        2.0              0         0.0     0.0                0   
2    1   48        1.0              1        20.0     0.0                0   
3    0   61        3.0              1        30.0     0.0                0   
4    0   46        3.0              1        23.0     0.0                0   

   prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  heartRate  glucose  \
0             0         0    195.0  106.0   70.0  26.97       80.0     77.0   
1             0         0    250.0  121.0   81.0  28.73       95.0     76.0   
2             0         0    245.0  127.5   80.0  25.34       75.0     70.0   
3             1         0    225.0  150.0   95.0  28.58       65.0    103.0   
4             0         0    285.0  130.0   84.0  23.10       85.0     85.0   

   TenYearCHD  
0           0  
1           0  
2           0  
3           1  
4           0

# Exploratory Data Analysis

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4238 entries, 0 to 4237
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   sex              4238 non-null   int64  
 1   age              4238 non-null   int64  
 2   education        4133 non-null   float64
 3   currentSmoker    4238 non-null   int64  
 4   cigsPerDay       4209 non-null   float64
 5   BPMeds           4185 non-null   float64
 6   prevalentStroke  4238 non-null   int64  
 7   prevalentHyp     4238 non-null   int64  
 8   diabetes         4238 non-null   int64  
 9   totChol          4188 non-null   float64
 10  sysBP            4238 non-null   float64
 11  diaBP            4238 non-null   float64
 12  BMI              4219 non-null   float64
 13  heartRate        4237 non-null   float64
 14  glucose          3850 non-null   float64
 15  TenYearCHD       4238 non-null   int64  
dtypes: float64(9), int64(7)
memory usage: 529.8 KB


In [4]:
df.describe()

sex          age    education  currentSmoker   cigsPerDay  \
count  4238.000000  4238.000000  4133.000000    4238.000000  4209.000000   
mean      0.429212    49.584946     1.978950       0.494101     9.003089   
std       0.495022     8.572160     1.019791       0.500024    11.920094   
min       0.000000    32.000000     1.000000       0.000000     0.000000   
25%       0.000000    42.000000     1.000000       0.000000     0.000000   
50%       0.000000    49.000000     2.000000       0.000000     0.000000   
75%       1.000000    56.000000     3.000000       1.000000    20.000000   
max       1.000000    70.000000     4.000000       1.000000    70.000000   

            BPMeds  prevalentStroke  prevalentHyp     diabetes      totChol  \
count  4185.000000      4238.000000   4238.000000  4238.000000  4188.000000   
mean      0.029630         0.005899      0.310524     0.025720   236.721585   
std       0.169584         0.076587      0.462763     0.158316    44.590334   
min       0.000000         0.000000      0.000000     0.000000   107.000000   
25%       0.000000         0.000000      0.000000     0.000000   206.000000   
50%       0.000000         0.000000      0.000000     0.000000   234.000000   
75%       0.000000         0.000000      1.000000     0.000000   263.000000   
max       1.000000         1.000000      1.000000     1.000000   696.000000   

             sysBP        diaBP          BMI    heartRate      glucose  \
count  4238.000000  4238.000000  4219.000000  4237.000000  3850.000000   
mean    132.352407    82.893464    25.802008    75.878924    81.966753   
std      22.038097    11.910850     4.080111    12.026596    23.959998   
min      83.500000    48.000000    15.540000    44.000000    40.000000   
25%     117.000000    75.000000    23.070000    68.000000    71.000000   
50%     128.000000    82.000000    25.400000    75.000000    78.000000   
75%     144.000000    89.875000    28.040000    83.000000    87.000000   
max     295.000000   142.500000    56.800000   143.000000   394.000000   

        TenYearCHD  
count  4238.000000  
mean      0.151958  
std       0.359023  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       0.000000  
max       1.000000

In [5]:
df.isnull().sum()

sex                  0
age                  0
education          105
currentSmoker        0
cigsPerDay          29
BPMeds              53
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol             50
sysBP                0
diaBP                0
BMI                 19
heartRate            1
glucose            388
TenYearCHD           0
dtype: int64

In [6]:
df.columns

Index(['sex', 'age', 'education', 'currentSmoker', 'cigsPerDay', 'BPMeds',
       'prevalentStroke', 'prevalentHyp', 'diabetes', 'totChol', 'sysBP',
       'diaBP', 'BMI', 'heartRate', 'glucose', 'TenYearCHD'],
      dtype='object')

In [7]:
#sns.pairplot(df)

In [8]:
#for assumption checking---normality
#sns.distplot(df['TenYearCHD'])

In [9]:
#import seaborn as sns
#fig,ax=plt.subplots(figsize=(14,12))
#sns.heatmap(df.corr(),annot=True)

# For fill out missing values variable now we fit model on that variables using other available complete data/variables

Complete data: variables
sex , age  ,currentSmoker ,prevalentStroke ,prevalentHyp , diabetes ,sysBP , diaBP  ,TenYearCHD 


Missing data present : variables
education ,cigsPerDay ,BPMeds ,totChol  ,BMI , heartRate ,glucose  

In [7]:
l=['education', 'cigsPerDay','totChol','BMI', 'heartRate', 'glucose']

for i in l:
    d=df[df[i].isnull()].index.tolist()

    df1=df.drop(df.index[[d]])
    
    #fit regression model
    
    y=df1[i]
    x=df1[['sex', 'age','currentSmoker','prevalentStroke', 'prevalentHyp', 'diabetes','sysBP','diaBP','TenYearCHD']]
    
    from sklearn.model_selection import train_test_split

    x_tr,x_te,y_tr,y_te=train_test_split(x,y,test_size=0.3,random_state=101)
    
    from sklearn.linear_model import LinearRegression

    lm=LinearRegression()

    lm.fit(x_tr,y_tr) 
    
    print("Fitted model is given as :",i," = ",round(lm.intercept_,4),"+ ",round(lm.coef_[0],4),"sex + ",
          round(lm.coef_[1],4),"age  + ",round(lm.coef_[2],4),"currentSmoker  + ",round(lm.coef_[3],4),"prevalentStroke  + ",
          round(lm.coef_[4],4),"prevalentHyp  + ",round(lm.coef_[5],4),"diabetes  +",round(lm.coef_[6],4),"sysBP  + ",
          round(lm.coef_[6],4)," diaBP  +",round(lm.coef_[8],4)," TenYearCHD")

    # Prediction
    y_pred1=lm.predict(x_te);
    #y_pred1
    
    # For fill out missing value in Age variable now we fit model on Age using other available complete data
    
    df[i]=np.where(df[i].isnull(),round(lm.intercept_,4)+ round(lm.coef_[0],4)* df['sex']+round(lm.coef_[1],4)*df['age']+
                   round(lm.coef_[2],4)*df['currentSmoker']+round(lm.coef_[3],4)*df['prevalentStroke']+round(lm.coef_[4],4)*df['prevalentHyp']+ 
                   round(lm.coef_[5],4)* df['diabetes']+round(lm.coef_[6],4)*df['sysBP']+round(lm.coef_[7],4)*df['diaBP']+
                   round(lm.coef_[8],4)*df['TenYearCHD'],df[i])
    


C:\Users\DeLL\Anaconda3\lib\site-packages\pandas\core\indexes\base.py:4307: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  result = getitem(key)


Fitted model is given as : education  =  3.2279 +  0.0467 sex +  -0.015 age  +  -0.0675 currentSmoker  +  -0.3832 prevalentStroke  +  0.0743 prevalentHyp  +  -0.1416 diabetes  + -0.0063 sysBP  +  -0.0063  diaBP  + -0.0135  TenYearCHD
Fitted model is given as : cigsPerDay  =  0.1511 +  4.0906 sex +  -0.0623 age  +  17.6097 currentSmoker  +  -0.8073 prevalentStroke  +  0.0653 prevalentHyp  +  -0.8889 diabetes  + 0.0113 sysBP  +  0.0113  diaBP  + 0.7216  TenYearCHD
Fitted model is given as : totChol  =  136.1826 +  -6.3567 sex +  1.1405 age  +  3.5443 currentSmoker  +  -11.1036 prevalentStroke  +  0.1643 prevalentHyp  +  -1.8651 diabetes  + 0.0951 sysBP  +  0.0951  diaBP  + 0.6637  TenYearCHD
Fitted model is given as : BMI  =  16.6344 +  0.7569 sex +  0.0109 age  +  -1.0435 currentSmoker  +  -0.4288 prevalentStroke  +  0.8308 prevalentHyp  +  1.3445 diabetes  + -0.005 sysBP  +  -0.005  diaBP  + -0.085  TenYearCHD
Fitted model is given as : heartRate  =  65.0047 +  -3.2778 sex +  -0.0805 a

C:\Users\DeLL\Anaconda3\lib\site-packages\pandas\core\indexes\base.py:4307: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  result = getitem(key)
C:\Users\DeLL\Anaconda3\lib\site-packages\pandas\core\indexes\base.py:4307: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  result = getitem(key)
C:\Users\DeLL\Anaconda3\lib\site-packages\pandas\core\indexes\base.py:4307: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an arr

Fitted model is given as : glucose  =  69.3471 +  0.246 sex +  0.0359 age  +  -1.1458 currentSmoker  +  -2.3774 prevalentStroke  +  -0.8743 prevalentHyp  +  83.4746 diabetes  + 0.1041 sysBP  +  0.1041  diaBP  + 5.093  TenYearCHD


In [11]:
df.isnull().sum()

sex                 0
age                 0
education           0
currentSmoker       0
cigsPerDay          0
BPMeds             53
prevalentStroke     0
prevalentHyp        0
diabetes            0
totChol             0
sysBP               0
diaBP               0
BMI                 0
heartRate           0
glucose             0
TenYearCHD          0
dtype: int64

In [8]:
z=df[df['BPMeds'].isnull()].index.tolist()

df2=df.drop(df.index[[z]])
df2

C:\Users\DeLL\Anaconda3\lib\site-packages\pandas\core\indexes\base.py:4307: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  result = getitem(key)


sex  age  education  currentSmoker  cigsPerDay  BPMeds  prevalentStroke  \
0       1   39        4.0              0         0.0     0.0                0   
1       0   46        2.0              0         0.0     0.0                0   
2       1   48        1.0              1        20.0     0.0                0   
3       0   61        3.0              1        30.0     0.0                0   
4       0   46        3.0              1        23.0     0.0                0   
...   ...  ...        ...            ...         ...     ...              ...   
4232    1   68        1.0              0         0.0     0.0                0   
4233    1   50        1.0              1         1.0     0.0                0   
4234    1   51        3.0              1        43.0     0.0                0   
4236    0   44        1.0              1        15.0     0.0                0   
4237    0   52        2.0              0         0.0     0.0                0   

      prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  heartRate  \
0                0         0    195.0  106.0   70.0  26.97       80.0   
1                0         0    250.0  121.0   81.0  28.73       95.0   
2                0         0    245.0  127.5   80.0  25.34       75.0   
3                1         0    225.0  150.0   95.0  28.58       65.0   
4                0         0    285.0  130.0   84.0  23.10       85.0   
...            ...       ...      ...    ...    ...    ...        ...   
4232             1         0    176.0  168.0   97.0  23.14       60.0   
4233             1         0    313.0  179.0   92.0  25.97       66.0   
4234             0         0    207.0  126.5   80.0  19.71       65.0   
4236             0         0    210.0  126.5   87.0  19.16       86.0   
4237             0         0    269.0  133.5   83.0  21.47       80.0   

        glucose  TenYearCHD  
0      77.00000           0  
1      76.00000           0  
2      70.00000           0  
3     103.00000           1  
4      85.00000           0  
...         ...         ...  
4232   79.00000           1  
4233   86.00000           1  
4234   68.00000           0  
4236   77.22495           0  
4237  107.00000           0  

[4185 rows x 16 columns]

a=df2[['BPMeds']]
a.astype(int)

#df2=df2.append('BPMeds_')
#df2

In [10]:
y2=round(df2[['BPMeds']])
x2=df2[['sex', 'age','currentSmoker','prevalentStroke', 'prevalentHyp', 'diabetes','sysBP','diaBP','TenYearCHD']]
    

    
    
from sklearn.model_selection import train_test_split

x_tr,x_te,y_tr,y_te=train_test_split(x2,y2,test_size=0.3,random_state=101)
    

    
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

lm=LogisticRegression(solver='liblinear',random_state=0).fit(x_tr,y_tr)
    

    
print("Fitted model is given as :  BPMeds = ",round(lm.intercept_[0],4),"+ ",round(lm.coef_[0,0],4),"sex + ",
      round(lm.coef_[0,1],4),"age  + ",round(lm.coef_[0,2],4),"currentSmoker  + ",
      round(lm.coef_[0,3],4),"prevalentStroke  + ",round(lm.coef_[0,4],4),"prevalentHyp  + ",
      round(lm.coef_[0,5],4),"diabetes  +",round(lm.coef_[0,6],4),"sysBP  + ",round(lm.coef_[0,6],4)," diaBP  +",
      round(lm.coef_[0,8],4)," TenYearCHD")



# Prediction
y_pred2=lm.predict(x_te);
print(y_pred2)

    
# For fill out missing value in Age variable now we fit model on Age using other available complete data
    
df['BPMeds']=np.where(df['BPMeds'].isnull(),round(lm.intercept_[0],4)+ round(lm.coef_[0,0],4)* df['sex']+
                      round(lm.coef_[0,1],4)*df['age']+round(lm.coef_[0,2],4)*df['currentSmoker']+
                      round(lm.coef_[0,3],4)*df['prevalentStroke']+round(lm.coef_[0,4],4)*df['prevalentHyp']+ 
                      round(lm.coef_[0,5],4)* df['diabetes']+round(lm.coef_[0,6],4)*df['sysBP']+
                      round(lm.coef_[0,7],4)*df['diaBP']+round(lm.coef_[0,8],4)*df['TenYearCHD'],df['BPMeds'])
    

Fitted model is given as :  BPMeds =  -3.8427 +  -0.5651 sex +  -0.0366 age  +  -0.3318 currentSmoker  +  1.3562 prevalentStroke  +  3.3043 prevalentHyp  +  0.0805 diabetes  + 0.0156 sysBP  +  0.0156  diaBP  + 0.6951  TenYearCHD
[0. 0. 0. ... 0. 0. 0.]


C:\Users\DeLL\Anaconda3\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


In [14]:
df.isnull().sum()

sex                0
age                0
education          0
currentSmoker      0
cigsPerDay         0
BPMeds             0
prevalentStroke    0
prevalentHyp       0
diabetes           0
totChol            0
sysBP              0
diaBP              0
BMI                0
heartRate          0
glucose            0
TenYearCHD         0
dtype: int64

Now , data contains no any missing value in any variable.
Now , we will obtain a final model

In [15]:
y_=df['TenYearCHD']
x_=df[['sex', 'age', 'education', 'currentSmoker', 'cigsPerDay', 'BPMeds',
       'prevalentStroke', 'prevalentHyp', 'diabetes', 'totChol', 'sysBP',
       'diaBP', 'BMI', 'heartRate', 'glucose']]
np.shape(x_)

(4238, 15)

In [16]:
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test=train_test_split(x_,y_,test_size=0.3,random_state=101)
print(x_train.shape)
print(x_test.shape)

(2966, 15)
(1272, 15)


# Evaluating  using knn classification

In [17]:
from sklearn.neighbors import KNeighborsClassifier

knn=KNeighborsClassifier(n_neighbors=25)

In [18]:
knn.fit(x_train,y_train)

KNeighborsClassifier(n_neighbors=25)

In [19]:
#Predict the respose for test dataset

y_pred_=knn.predict(x_test)
y_pred_

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [20]:
#Confusion matrix

from sklearn.metrics import confusion_matrix

confMat=confusion_matrix(y_test,y_pred_)
confMat

array([[1093,    4],
       [ 170,    5]], dtype=int64)

In [21]:
#import scikit-learn metrics module for aaccuracy calculation
from sklearn import metrics

from sklearn.metrics import accuracy_score
#model accuracy

print("Accuracy:",metrics.accuracy_score(y_test,y_pred_))

Accuracy: 0.8632075471698113


Thus this model explained 86.32 % variability in 10 year risk of coronary heart disease CHD when missing values are replace using linear regression and logistic regression respective to that particular column and variable type.

# Evaluating  using logistic regression

In [22]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

lm=LogisticRegression(solver='liblinear',random_state=101).fit(x_train,y_train)


In [24]:
print("R square : ", lm.score(x_train,y_train))

R square :  0.8509777478084963
